### ISSUE TITLE
Memory Leak Media Scanner Connection

### ISSUE THREAD

**Issue by renegens** [miguelbcr/RxPaparazzo#58]
Has anybody encountered that LeakCanary is complaining about memory leaks? This happens when rxPaparazzo is used on an Activity, when used on fragment I don't see any leaks.  Below is what I am using the take either a picture or retrieve from gallery. 

```
compositeSubscription.add(RxPaparazzo.takeImage(this)
                   .size(new CustomMaxSize(900))
                   .usingGallery()
                   .subscribeOn(Schedulers.io())
                   .observeOn(AndroidSchedulers.mainThread())
                   .doOnError(Timber::d)
                   .subscribe(response -> {
                       if (response.resultCode() != RESULT_OK) {
                           response.targetUI()
                                   .showUserCanceled();
                           return;
                       }
                       response.targetUI()
                               .loadPicture(response.data());
                   }));
```

**Comment by JamesMcIntosh** [comment]
A quick Google pulls up this commit in LeakCanary, it says that there is a problem with the MediaScannerConnection.scanFile() which fails to clean up the reference to the Context by not disconnecting properly. I'm not sure why they have version limited the bug as it's not fixed.

https://github.com/square/leakcanary/pull/111/files
https://code.google.com/p/android/issues/detail?id=173788

The suggested workaround is to self-manage the MediaScannerConnection.

**Comment by JamesMcIntosh** [comment]
After a little more reading I've changed it to use the Intent.ACTION_MEDIA_SCANNER_SCAN_FILE on the PR since we don't require any callback.
https://github.com/miguelbcr/RxPaparazzo/pull/53/commits/4aa9517c40618e4f774bb8e30ca98b81f160e772

**Comment by renegens** [comment]
@JamesMcIntosh Have you confirmed that this stops the leak?

**Comment by JamesMcIntosh** [comment]
Using an Intent stops the leak because the App no longer directly interacts with the media scanner

**Comment by renegens** [comment]
@miguelbcr could you clarify if this will be added in the next release?

**Comment by miguelbcr** [comment]
Hi @renegens,
yes. it will be added

**Comment by miguelbcr** [comment]
Hi @renegens ,
we've released [version 0.5.1-2x](https://github.com/miguelbcr/RxPaparazzo/releases/tag/0.5.1-2x) which solves this issue.
Note that this new version add picking file functionality and API has been redesigned, so take a look to [README.md](https://github.com/miguelbcr/RxPaparazzo/blob/2.x/README.md)
Let us know if it works for you.

**Comment by renegens** [comment]
Hello @miguelbcr 

Thanks for getting back to me. I am unable to check if is working  beacuse I am on RxJava v1 and I am getting some weired conflicts. Will you add a v1 version of this?

**Comment by miguelbcr** [comment]
Hi @renegens,
I've released [0.4.7 version](https://github.com/miguelbcr/RxPaparazzo/releases/tag/0.4.7) which solves this issue.
Check it out.

By the way, I don't know if you are aware of [RxJava2Interop](https://github.com/akarnokd/RxJava2Interop) which allows convert between RxJava 1.x and 2.x reactive types, so you could use both RxJava 1.x and 2.x on the same chain.

**Comment by renegens** [comment]
@miguelbcr thanks for acting so quick. The initial leak went away but I am getting still a leack. I am attaching the leak canary analysis. The question now is again, who's code is problematic :P. 

Btw I saw the Interpro sounds good. So you basically add  RxJava 2 and although you have rxjava observables you can subscribe to them in RxJava 2 manner and wrap them in disposables?

`In com.globyworks.citykey.debug:2.1.11-DEBUG:111.
* com.globyworks.citykey.ui.userprofile.editprofile.EditProfileActivity has leaked:
* GC ROOT static rx_activity_result.HolderActivity.request
* references rx_activity_result.Request.onPreResult
* references com.miguelbcr.ui.rx_paparazzo.interactors.PickImage$2.this$0 (anonymous implementation of rx_activity_result.OnPreResult)
* references com.miguelbcr.ui.rx_paparazzo.interactors.PickImage.getPath
* references com.miguelbcr.ui.rx_paparazzo.interactors.GetPath.targetUi
* references com.miguelbcr.ui.rx_paparazzo.entities.TargetUi.ui
* leaks com.globyworks.citykey.ui.userprofile.editprofile.EditProfileActivity instance

* Retaining: 99 KB.
* Reference Key: 00402425-76c9-413c-800e-2848d0e4319f
* Device: motorola google Nexus 6 shamu
* Android Version: 7.1.1 API: 25 LeakCanary: 1.5 00f37f5
* Durations: watch=75653ms, gc=174ms, heap dump=9996ms, analysis=83250ms

* Details:
* Class rx_activity_result.HolderActivity
|   static request = rx_activity_result.Request@320517648 (0x131ab610)
|   static $classOverhead = byte[1800]@315291649 (0x12caf801)
* Instance of rx_activity_result.Request
|   static $classOverhead = byte[72]@315870401 (0x12d3ccc1)
|   intent = android.content.Intent@320524520 (0x131ad0e8)
|   onPreResult = com.miguelbcr.ui.rx_paparazzo.interactors.PickImage$2@320480112 (0x131a2370)
|   onResult = rx_activity_result.RxActivityResult$Builder$2@320480160 (0x131a23a0)
|   shadow$_klass_ = rx_activity_result.Request
|   shadow$_monitor_ = 0
* Instance of com.miguelbcr.ui.rx_paparazzo.interactors.PickImage$2
|   static $classOverhead = byte[56]@315661185 (0x12d09b81)
|   this$0 = com.miguelbcr.ui.rx_paparazzo.interactors.PickImage@320431104 (0x13196400)
|   shadow$_klass_ = com.miguelbcr.ui.rx_paparazzo.interactors.PickImage$2
|   shadow$_monitor_ = 0
* Instance of com.miguelbcr.ui.rx_paparazzo.interactors.PickImage
|   static $classOverhead = byte[56]@315661377 (0x12d09c41)
|   getPath = com.miguelbcr.ui.rx_paparazzo.interactors.GetPath@320401312 (0x1318efa0)
|   startIntent = com.miguelbcr.ui.rx_paparazzo.interactors.StartIntent@320401264 (0x1318ef70)
|   shadow$_klass_ = com.miguelbcr.ui.rx_paparazzo.interactors.PickImage
|   shadow$_monitor_ = 0
* Instance of com.miguelbcr.ui.rx_paparazzo.interactors.GetPath
|   static $classOverhead = byte[60]@315659649 (0x12d09581)
|   downloadImage = com.miguelbcr.ui.rx_paparazzo.interactors.DownloadImage@320401288 (0x1318ef88)
|   targetUi = com.miguelbcr.ui.rx_paparazzo.entities.TargetUi@320431056 (0x131963d0)
|   uri = android.net.Uri$HierarchicalUri@321014944 (0x13224ca0)
|   shadow$_klass_ = com.miguelbcr.ui.rx_paparazzo.interactors.GetPath
|   shadow$_monitor_ = 0
* Instance of com.miguelbcr.ui.rx_paparazzo.entities.TargetUi
|   static $classOverhead = byte[72]@315653313 (0x12d07cc1)
|   ui = com.globyworks.citykey.ui.userprofile.editprofile.EditProfileActivity@320772800 (0x131e9ac0)
|   shadow$_klass_ = com.miguelbcr.ui.rx_paparazzo.entities.TargetUi
|   shadow$_monitor_ = 0
* Instance of com.globyworks.citykey.ui.userprofile.editprofile.EditProfileActivity
|   static $classOverhead = byte[2008]@319086593 (0x1304e001)
|   static ajc$tjp_0 = org.aspectj.runtime.reflect.JoinPointImpl$StaticPartImpl@320498680 (0x131a6bf8)
|   changePicture = android.support.v7.widget.AppCompatButton@319544320 (0x130bdc00)
|   changeUserPassword = android.support.v7.widget.AppCompatButton@319581184 (0x130c6c00)
|   compositeSubscription = rx.subscriptions.CompositeSubscription@320790976 (0x131ee1c0)
|   presenter = com.globyworks.citykey.ui.userprofile.editprofile.EditProfilePresenter@320695912 (0x131d6e68)
|   progressDialog = null
|   rootView = android.widget.LinearLayout@319533056 (0x130bb000)
|   userEmail = android.support.design.widget.TextInputEditText@319577088 (0x130c5c00)
|   userImage = de.hdodenhof.circleimageview.CircleImageView@319543296 (0x130bd800)
|   userName = android.support.design.widget.TextInputEditText@319548416 (0x130bec00)
|   userPhone = android.support.design.widget.TextInputEditText@319580160 (0x130c6800)
|   userViewModel = com.globyworks.citykey.models.viewmodels.UserViewModel@319452832 (0x130a76a0)
|   isOnline = true
|   presenter = com.globyworks.citykey.ui.userprofile.editprofile.EditProfilePresenter@320695912 (0x131d6e68)
|   rootView = android.widget.LinearLayout@319533056 (0x130bb000)
|   snackbar = android.support.design.widget.Snackbar@320691384 (0x131d5cb8)
|   subscription = rx.observers.SafeSubscriber@320851688 (0x131fcee8)
|   toolbar = android.support.v7.widget.Toolbar@319535104 (0x130bb800)
|   mDelegate = android.support.v7.app.AppCompatDelegateImplN@319248272 (0x13075790)
|   mEatKeyUpEvent = false
|   mResources = null
|   mThemeId = 2131492932
|   mCreated = true
|   mFragments = android.support.v4.app.FragmentController@319292528 (0x13080470)
|   mHandler = android.support.v4.app.FragmentActivity$1@321039616 (0x1322ad00)
|   mNextCandidateRequestIndex = 0
|   mOptionsMenuInvalidated = false
|   mPendingFragmentActivityResults = android.support.v4.util.SparseArrayCompat@320498008 (0x131a6958)
|   mReallyStopped = true
|   mRequestedPermissionsFromFragment = false
|   mResumed = false
|   mRetaining = false
|   mStopped = true
|   mStartedActivityFromFragment = false
|   mStartedIntentSenderFromFragment = false
|   mExtraDataMap = android.support.v4.util.SimpleArrayMap@320498584 (0x131a6b98)
|   mActionBar = null
|   mActionModeTypeStarting = 0
|   mActivityInfo = android.content.pm.ActivityInfo@319248416 (0x13075820)
|   mActivityTransitionState = android.app.ActivityTransitionState@320080016 (0x13140890)
|   mApplication = com.globyworks.citykey.root.CityKeyApplication@315071136 (0x12c79aa0)
|   mCalled = true
|   mChangeCanvasToTranslucent = false
|   mChangingConfigurations = false
|   mComponent = android.content.ComponentName@320536720 (0x131b0090)
|   mConfigChangeFlags = 0
|   mCurrentConfig = android.content.res.Configuration@319500640 (0x130b3160)
|   mDecor = null
|   mDefaultKeyMode = 0
|   mDefaultKeySsb = null
|   mDestroyed = true
|   mDoReportFullyDrawn = false
|   mEatKeyUpEvent = false
|   mEmbeddedID = null
|   mEnableDefaultActionBarUp = true
|   mEnterTransitionListener = android.app.SharedElementCallback$1@1867084992 (0x6f4974c0)
|   mExitTransitionListener = android.app.SharedElementCallback$1@1867084992 (0x6f4974c0)
|   mFinished = true
|   mFragments = android.app.FragmentController@319292576 (0x130804a0)
|   mHandler = android.os.Handler@321040160 (0x1322af20)
|   mHasCurrentPermissionsRequest = false
|   mIdent = 85366095
|   mInstanceTracker = android.os.StrictMode$InstanceTracker@319292560 (0x13080490)
|   mInstrumentation = android.app.Instrumentation@316793688 (0x12e1e358)
|   mIntent = android.content.Intent@320995272 (0x1321ffc8)
|   mLastNonConfigurationInstances = null
|   mMainThread = android.app.ActivityThread@314585344 (0x12c03100)
|   mManagedCursors = java.util.ArrayList@320498632 (0x131a6bc8)
|   mManagedDialogs = null
|   mMenuInflater = null
|   mParent = null
|   mReferrer = java.lang.String@320732392 (0x131dfce8)
|   mResultCode = 0
|   mResultData = null
|   mResumed = false
|   mSearchEvent = null
|   mSearchManager = null
|   mStartedActivity = false
|   mStopped = true
|   mTaskDescription = android.app.ActivityManager$TaskDescription@321039648 (0x1322ad20)
|   mTemporaryPause = false
|   mTitle = java.lang.String@320421248 (0x13193d80)
|   mTitleColor = 0
|   mTitleReady = true
|   mToken = android.os.BinderProxy@320814976 (0x131f3f80)
|   mTranslucentCallback = null
|   mUiThread = java.lang.Thread@1936614928 (0x736e6610)
|   mVisibleBehind = false
|   mVisibleFromClient = true
|   mVisibleFromServer = true
|   mVoiceInteractor = null
|   mWindow = com.android.internal.policy.PhoneWindow@321006912 (0x13222d40)
|   mWindowAdded = true
|   mWindowManager = android.view.WindowManagerImpl@320498272 (0x131a6a60)
|   mInflater = com.android.internal.policy.PhoneLayoutInflater@319480160 (0x130ae160)
|   mOverrideConfiguration = null
|   mResources = android.content.res.Resources@320421368 (0x13193df8)
|   mTheme = android.content.res.Resources$Theme@319292384 (0x130803e0)
|   mThemeResource = 2131492932
|   mBase = android.app.ContextImpl@320736720 (0x131e0dd0)
|   shadow$_klass_ = com.globyworks.citykey.ui.userprofile.editprofile.EditProfileActivity
|   shadow$_monitor_ = 1242818348
* Excluded Refs:
| Field: android.view.Choreographer$FrameDisplayEventReceiver.mMessageQueue (always)
| Thread:FinalizerWatchdogDaemon (always)
| Thread:main (always)
| Thread:LeakCanary-Heap-Dump (always)
| Class:java.lang.ref.WeakReference (always)
| Class:java.lang.ref.SoftReference (always)
| Class:java.lang.ref.PhantomReference (always)
| Class:java.lang.ref.Finalizer (always)
| Class:java.lang.ref.FinalizerReference (always)`

**Comment by miguelbcr** [comment]
Hi @renegens,
Regarding to your question about RxJava2Interop, I think so, but sincerely, I've only used it once.

And regarding the leak it could be related with [getPath](https://github.com/miguelbcr/RxPaparazzo/blob/master/rx_paparazzo/src/main/java/com/miguelbcr/ui/rx_paparazzo/interactors/PickImage.java#L64) instance that keeps a instance of [targetUi](https://github.com/miguelbcr/RxPaparazzo/blob/master/rx_paparazzo/src/main/java/com/miguelbcr/ui/rx_paparazzo/interactors/GetPath.java#L32)

Anyway I have no maintenance plans for 1.x branch, so you could try to use 2.x branch because it has been refactorized and maybe the leak has been solved, but I can't assure you.

Also you could try 1.x sample app (I've added leakCanary lib) and observe if the leak remains (I've been testing but no leaks have appeared)

Let me know if you figure out something.

**Comment by renegens** [comment]
Yes you are correct. I believe that the memory leak is not related to this library but actually to RxActivity which has a static reference to the HolderActivity which has the response.  Apart from that, I have trouble adding the v2 library of paparazzo. Any thoughts on that or should I open a new issue?

<img width="497" alt="screen shot 2017-03-01 at 23 16 44" src="https://cloud.githubusercontent.com/assets/8403585/23482039/4d75cfe6-fed6-11e6-8a61-f9af1dd16ac3.png">

**Comment by JamesMcIntosh** [comment]
Which repository are you trying to resolve it from?

If you use JitPack it is available - https://jitpack.io/#miguelbcr/RxPaparazzo

If you are building the branch locally and running ./gradlew rx_paparazzo:install to publish into your local maven repository there is a small bug, the version number is "2.x-agrimap", I will make a PR to fix that.

**Comment by JamesMcIntosh** [comment]
I see the problem now. In the docs the version is 0.5.1-2.x where the actual published version is 0.5.1-2x

**Comment by renegens** [comment]
@JamesMcIntosh yeap that's it. I think I need new glasses :P or just call it a day :)

**Comment by miguelbcr** [comment]
Sorry, it was my mistake... :)
I've released [0.5.2-2.x](https://github.com/miguelbcr/RxPaparazzo/releases/tag/0.5.2-2.x)

**Comment by renegens** [comment]
Thanks for all the help and work @JamesMcIntosh, @miguelbcr
